### DISCUSSION TITLE
JsWriter - Compile time compute as much as possible?

### DISCUSSION THREAD

**Discussion by dsherret** [swc-project/swc#9003] [General]
I was looking at the code for `WriteJs`/`JsWriter` and I believe we could use a proc macro to turn this:

```rs
keyword!("import")
```

into...

```rs
{
  const VALUE = PreComputedText {
    text: "import",
    line_count: 0,
    last_line_char_count: 6,
  };
  self.wr.write_precomputed(&VALUE);
}
```

This would avoid having to do any calculations for stuff at runtime that's known at compile time and improve the source map emit's performance.

**Comment by kdy1** [comment]
Can you profile it? I think rustc will optimize it
